workers/unity/Assets/Gdk/Core/Utility/ConnectionUtility.cs

```diff
@@ -89,45 +75,67 @@ private static ConnectionParameters CreateConnectionParameters(ConnectionConfig
             return connectionParameters;
         }
 
-        private static Connection TryToConnect(Future<Connection> connectionFuture)
+        private static Connection TryToConnect(Future<Connection> connectionFuture, out ConnectionError error)
         {
             var connection = connectionFuture.Get(RuntimeConfigDefaults.ConnectionTimeout);
 
             if (!connection.HasValue || !connection.Value.IsConnected)
             {
-                Debug.LogError("Failed to connect to SpatialOS.");
+                error = new ConnectionError
+                {
+                    Message = "Failed to connect to SpatialOS.",
+                    Reason = ConnectionErrorReseaon.CannotEstablishConnection
+                };
+
                 return null;
             }
 
             Debug.Log("Successfully connected to SpatialOS!");
+            error = default(ConnectionError);
 
             return connection.Value;
         }
 
-        private static string GetDeploymentName(Locator locator)
+        private static string GetDeploymentName(Locator locator, out ConnectionError error)
         {
             using (var deploymentsFuture = locator.GetDeploymentListAsync())
             {
                 var deployments = deploymentsFuture.Get(RuntimeConfigDefaults.ConnectionTimeout);
 
                 if (!deployments.HasValue)
                 {
-                    Debug.LogError("Failed to retrieve deployment.");
+                    error = new ConnectionError
+                    {
+                        Message = "Failed to retrieve deployment.",
+                        Reason = ConnectionErrorReseaon.DeploymentNotFound
+                    };
+
                     return null;
                 }
 
                 if (deployments.Value.Error != null)
                 {
-                    Debug.LogError($"Failed to obtain deployment name with error: {deployments.Value.Error}.");
+                    error = new ConnectionError
+                    {
+                        Message = $"Failed to obtain deployment name with error: {deployments.Value.Error}.",
+                        Reason = ConnectionErrorReseaon.DeploymentNotFound
+                    };
+
                     return null;
                 }
 
                 if (deployments.Value.Deployments.Count == 0)
                 {
-                    Debug.LogError("Received an empty list of deployments.");
+                    error = new ConnectionError
+                    {
+                        Message = "Received an empty list of deployments.",
+                        Reason = ConnectionErrorReseaon.DeploymentNotFound
+                    };
+
                     return null;
                 }
 
+                error = default(ConnectionError);
```

Comment:
I don't think we should assign error any value, if there is no error

Reply:
If you pass an `out` parameter to a function it always has to be assigned. Even if this is not intuitive I think we have to keep it. Its similar to the implementation of TryGetValue.